text/0033-view-hierarch.md

```diff
@@ -0,0 +1,46 @@
+* Start Date: 2022-11-02
+* RFC Type: feature
+* RFC PR: [#33](https://github.com/getsentry/rfcs/pull/33)
+* RFC Status: draft
+* RFC Driver: -
+
+# Summary
+
+Allow SDKs to send View hierarch of the application UI.
+
+# Motivation
+
+Being able to see what the user was looking at is a good way to found out problems in the application. 
+This RFC proposes a way to send information from the UI state of an application, known as "view hierarch".
+
+# Proposal 
+
+The proposal is to create a new envelope item in the form of a json representation of the UI elements, like so: 
+
+```json
+{
+    "type": "element_class_name", 
+    "identifier": "element_identifier",
+    "children": [...], //An array of ui elements.
+    "width": 100, 
+    "height": 100,
+    "depth": 100, //if 3D
+    "x": 0,
+    "y": 1,
+    "z": 2, //if 3D.
+    "visible": true|false,
+    "alpha": 1,
+    "{extra_properties}": "{property value}" //adicional information by platform
+}
+```
+
+# Options Considered
+
+- We could send the same information as attachment, but this impact on attachment quota. And also we need to rely on file name to determined whether an attachment is a view hierarch or not, this is a problem we already have with [screenshots](https://develop.sentry.dev/sdk/features/#screenshots).
```

Comment:
IMO it’s completely fair to send these as attachements. It costs us money to store and use this information.

We can make this a new attachment type so we don’t have the same problem with filename (we can probably do the same with screenshots as well).

Reply:
Indeed, good idea @AbhiPrasad 
https://develop.sentry.dev/sdk/envelopes/#attachment
Either a new envelope item type or a new attachment type -> `attachment_type` in the spec.
I'd go this way if we wanna charge for this feature.

Reply:
Thought about this more, and I'd vote for an attachment with a different `attachment_type` for the following reasons:
- We'd can immediately start sending and using this data, no need for large Relay updates
- Less new code in the SDK vs. adding a new envelope item, so less hassle to maintain/test
- We don't have to touch outcomes/rate limiting/processing code at all
- We always can eventually migrate to a new envelope item type if we feel the need to, that option is still left open.

Reply:
The only pending point is what older versions of Relay do with unknown `attachment_type`, is going to be a back compatibility issue and we are fine with it? I guess the same would happen with a new envelope item anyway.

Reply:
[As per this logic](https://github.com/getsentry/relay/blob/f0d3edc54b63f4b75f6b8c50a488369d8b2275b0/relay-server/src/envelope.rs#L608-L618), if a relay encounters an unknown attachment type it sets it to the default `event.attachment`: https://github.com/getsentry/relay/blob/f0d3edc54b63f4b75f6b8c50a488369d8b2275b0/relay-server/src/envelope.rs#L372-L376

Reply:
I really like the idea, but im wondering if its possible to de-obsfucate information in the attachment.

Reply:
We can apply processing on an attachment - this is what we do with minidumps for example.

Reply:
I've also been trying to figure out the work to use this new envelope type, so I met yesterday with @JoshFerge for insight on how they handled a new envelope type for replay, and he echoes @AbhiPrasad's comments r.e a substantial amount of work required in Relay to support this (adding the new type to relay and adding a consumer being the biggest portions)

If we have these sending as attachments, our team could likely do something quick to render the attachment preview (otherwise the user just gets their JSON which isn't too helpful) before we get into the main project development